bin/propolis-server/src/lib/migrate/protocol.rs

```diff
@@ -0,0 +1,323 @@
+        write!(
+            f,
+            "{}",
+            match self {
+                Encoding::Ron => "ron",
+            }
+        )
+    }
+}
+
+impl FromStr for Encoding {
+    type Err = ProtocolParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "ron" => Ok(Encoding::Ron),
+            _ => Err(ProtocolParseError::InvalidEncoding(s.to_owned())),
+        }
+    }
+}
+
+/// A protocol selection.
+//
+// N.B. The ordering of fields in this struct matters! It ensures that the
+//      derived impls of PartialOrd and Ord compare versions before encodings.
+//      This ensures that the negotiation process always selects the latest
+//      version irrespective of whether it has the most preferable encoding.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+struct ProtocolParts {
+    version: u32,
+    encoding: Encoding,
+}
+
+impl ProtocolParts {
+    fn offer_string(&self) -> String {
+        format!(
+            "{PREFIX}{}{ENCODING_VERSION_SEPARATOR}{}",
+            self.encoding, self.version
+        )
+    }
+}
+
+impl From<Protocol> for ProtocolParts {
+    fn from(value: Protocol) -> Self {
+        match value {
+            Protocol::RonV0 => {
+                ProtocolParts { version: 0, encoding: Encoding::Ron }
+            }
+        }
+    }
+}
+
+impl FromStr for ProtocolParts {
+    type Err = ProtocolParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let (encoding, version) = s
+            .strip_prefix(PREFIX)
+            .ok_or_else(|| ProtocolParseError::InvalidPrefix(s.to_owned()))?
+            .split_once(ENCODING_VERSION_SEPARATOR)
+            .ok_or_else(|| {
+                ProtocolParseError::NoEncodingVersionSeparator(s.to_owned())
+            })?;
+
+        let encoding = Encoding::from_str(encoding)?;
+        let version = version.parse().map_err(|e: ParseIntError| {
+            ProtocolParseError::InvalidVersionNumber(
+                version.to_owned(),
+                e.to_string(),
+            )
+        })?;
+
+        Ok(ProtocolParts { encoding, version })
+    }
+}
+
+/// The set of protocol variants supported by this migration library, expressed
+/// as a set of parts.
+static PROTOCOL_PARTS: Lazy<BTreeSet<ProtocolParts>> = Lazy::new(|| {
+    let mut set = BTreeSet::new();
+    for p in enum_iterator::all::<Protocol>() {
+        let _added = set.insert(ProtocolParts::from(p));
+        assert!(_added);
+    }
+
+    set
+});
+
+/// Constructs a protocol offer string from a peekable protocol iterator.
+fn make_protocol_offers_from_parts<
+    T: std::iter::Iterator<Item = ProtocolParts>,
+>(
+    mut iter: Peekable<T>,
+) -> String {
+    let mut s = String::new();
+    while let Some(p) = iter.next() {
+        s.push_str(&p.offer_string());
+        if iter.peek().is_some() {
+            s.push(DELIMITER);
+        }
+    }
+
+    s
+}
```

Comment:
All the `BTreeSet` machinery seems a bit much for while we have only one (or two) valid versions, but I suppose that since it's written (and tested), there's no harm in leaving it as-is

Reply:
I reached for them to avoid having to worry about sorting collections in the right places (since `select_compatible_protocol` needs to iterate over the offer in descending version order), but now that the code's baked a little more I think we can just use plain Vecs. Tidied this as up as part of 17dbf47.